zk-token-sdk/src/sigma_proofs/mod.rs

```diff
@@ -25,15 +25,37 @@ pub mod pubkey_proof;
 pub mod zero_balance_proof;
 
 #[cfg(not(target_os = "solana"))]
-use {crate::errors::ProofVerificationError, curve25519_dalek::scalar::Scalar};
+use {
+    crate::errors::ProofVerificationError,
+    curve25519_dalek::{ristretto::CompressedRistretto, scalar::Scalar},
+};
 
+/// Deserializes an optional slice of bytes to a compressed Ristretto point.
+///
+/// This is a helper function for deserializing byte encodings of sigma proofs. It is designed to
+/// be used with `std::slice::Chunks`.
 #[cfg(not(target_os = "solana"))]
-fn canonical_scalar_from_slice(bytes: &[u8]) -> Result<Scalar, ProofVerificationError> {
-    if bytes.len() != 32 {
-        return Err(ProofVerificationError::Deserialization);
-    }
+fn ristretto_point_from_optional_slice(
+    optional_slice: Option<&[u8]>,
+) -> Result<CompressedRistretto, ProofVerificationError> {
+    let slice = optional_slice.ok_or(ProofVerificationError::Deserialization)?;
+    let point_bytes = slice[..32]
+        .try_into()
+        .map_err(|_| ProofVerificationError::Deserialization)?;
 
-    let scalar_bytes = bytes[..32]
+    Ok(CompressedRistretto::from_slice(point_bytes))
+}
+
+/// Deserializes an optional slice of bytes to a scalar.
+///
+/// This is a helper function for deserializing byte encodings of sigma proofs. It is designed to
+/// be used with `std::slice::Chunks`.
+#[cfg(not(target_os = "solana"))]
+fn canonical_scalar_from_optional_slice(
+    optional_slice: Option<&[u8]>,
+) -> Result<Scalar, ProofVerificationError> {
+    let slice = optional_slice.ok_or(ProofVerificationError::Deserialization)?;
+    let scalar_bytes = slice[..32]
         .try_into()
         .map_err(|_| ProofVerificationError::Deserialization)?;
 
```

Comment:
What if these were actually TryFrom definitions on CompressedRistretto and Scalar? Something like:
```
impl TryFrom<Option<&[u8]>> for CompressedRistretto {
    type Error = ProofVerificationError;

    fn try_from(maybe_slice: Option<&[u8]>) -> Result<Self, Self::Error> {
        maybe_slice
            .and_then(|slice| slice[..32].try_into().ok()) // not sure if this line could be written better
            .ok_or(ProofVerificationError::Deserialization)
    }
}
```

Then the call would look like:
```
let Y_0 = CompressedRistretto::try_from(chunks.next())?;
```

Seems a little cleaner to me. Wdyt?

Reply:
Yes this is a nice suggestion and I really like it. However, it seems like rust does not allow this since both the type `CompressedRistretto`/`Scalar` as well as the trait `TryInto` is defined outside the crate. I experimented with using a type wrapper around `CompressedRistretto`/`Scalar`, but it seemed like it was adding more complexity.

Reply:
One alternative that I did considered while I was working on this PR was to just use
```
if bytes.len() != PUBKEY_VALIDITY_PROOF_LEN {
    return Err(ProofVerificationError::Deserialization.into());
}

let Y = CompressedRistretto::from_slice(chunks.next().unwrap());
let z = canonical_scalar_from_slice(chunks.next().unwrap())?; // what we had previously
```
I added the helper functions because I didn't quite like the `unwrap`s even though it is safe as long as we have the length check beforehand (like we had originally). What do you think about this approach?

Reply:
> rust does not allow this since both the type `CompressedRistretto`/`Scalar` as well as the trait `TryInto` is defined outside the crate

Ah nuts. I forgot about that; I'm sorry for wasting your time.

Given that, I'm okay with the helpers.
I don't love the phrase "optional slice", but I haven't thought of anything better so let's go with those names. They're just private fns anyway. I will go back and make a couple code suggestions on the fns now.